src/client.js

```diff
@@ -4670,20 +4673,31 @@ function setupCallEventHandler(client) {
             callEventBuffer = [];
         }
     }
-    client.on("sync", onSync);
+    client.on("sync", evaluateEventBuffer);
 
     function onEvent(event) {
-        if (event.getType().indexOf("m.call.") !== 0) {
-            // not a call event
-            if (event.isBeingDecrypted() || event.isDecryptionFailure()) {
-                // not *yet* a call event, but might become one...
-                event.once("Event.decrypted", onEvent);
-            }
-            return;
+        // any call events or ones that might be once they're decrypted
+        if (event.getType().indexOf("m.call.") === 0 || event.isBeingDecrypted()) {
+            // queue up for processing once all events from this sync have been
+            // processed (see above).
+            callEventBuffer.push(event);
+        }
+
+        if (event.isBeingDecrypted() || event.isDecryptionFailure()) {
```

Comment:
why do we want to listen for decryption if that has already failed? Because it might succeed later when the keys are received?

Reply:
Yep, this happens fairly often if the to_device vent with the keys just lags behind the message itself